src/otx/algo/keypoint_detection/utils/keypoint_eval.py

```diff
@@ -0,0 +1,408 @@
+        raise ValueError(msg)
+
+    if heatmaps.ndim == 3:
+        num_keypoints, y_dim, x_dim = heatmaps.shape
+        batch_size = None
+        heatmaps_flatten = heatmaps.reshape(num_keypoints, -1)
+    else:
+        batch_size, num_keypoints, y_dim, x_dim = heatmaps.shape
+        heatmaps_flatten = heatmaps.reshape(batch_size * num_keypoints, -1)
+
+    y_locs, x_locs = np.unravel_index(np.argmax(heatmaps_flatten, axis=1), shape=(y_dim, x_dim))
+    locs = np.stack((x_locs, y_locs), axis=-1).astype(np.float32)
+    vals = np.amax(heatmaps_flatten, axis=1)
+    locs[vals <= 0.0] = -1
+
+    if batch_size:
+        locs = locs.reshape(batch_size, num_keypoints, 2)
+        vals = vals.reshape(batch_size, num_keypoints)
+
+    return locs, vals
+
+
+def _calc_distances(preds: np.ndarray, gts: np.ndarray, mask: np.ndarray, norm_factor: np.ndarray) -> np.ndarray:
+    """Calculate the normalized distances between preds and target.
+
+    Note:
+        - instance number: N
+        - keypoint number: K
+        - keypoint dimension: D (normally, D=2 or D=3)
+
+    Args:
+        preds (np.ndarray[N, K, D]): Predicted keypoint location.
+        gts (np.ndarray[N, K, D]): Groundtruth keypoint location.
+        mask (np.ndarray[N, K]): Visibility of the target. False for invisible
+            joints, and True for visible. Invisible joints will be ignored for
+            accuracy calculation.
+        norm_factor (np.ndarray[N, D]): Normalization factor.
+            Typical value is heatmap_size.
+
+    Returns:
+        np.ndarray[K, N]: The normalized distances. \
+            If target keypoints are missing, the distance is -1.
+    """
+    batch_size, num_keypoints, _ = preds.shape
+    # set mask=0 when norm_factor==0
+    _mask = mask.copy()
+    _mask[np.where((norm_factor == 0).sum(1))[0], :] = False
+
+    distances = np.full((batch_size, num_keypoints), -1, dtype=np.float32)
+    # handle invalid values
+    norm_factor[np.where(norm_factor <= 0)] = 1e6
+    distances[_mask] = np.linalg.norm(((preds - gts) / norm_factor[:, None, :])[_mask], axis=-1)
+    return distances.T
+
+
+def _distance_acc(distances: np.ndarray, thr: float = 0.5) -> float:
+    """Return the percentage below the distance threshold.
+
+    Ignore distances values with -1.
+
+    Note:
+        - instance number: N
+
+    Args:
+        distances (np.ndarray[N, ]): The normalized distances.
+        thr (float): Threshold of the distances.
+
+    Returns:
+        float: Percentage of distances below the threshold. \
+            If all target keypoints are missing, return -1.
+    """
+    distance_valid = distances != -1
+    num_distance_valid = distance_valid.sum()
+    if num_distance_valid > 0:
+        return (distances[distance_valid] < thr).sum() / num_distance_valid
+    return -1
+
+
+def keypoint_pck_accuracy(
+    pred: np.ndarray,
+    gt: np.ndarray,
+    mask: np.ndarray,
+    thr: np.ndarray,
+    norm_factor: np.ndarray,
+) -> tuple:
```

Comment:
```suggestion
) -> tuple[np.ndarray, float, int]:
```
It seems that we can give move detail info

Reply:
fixed at 64b81f26eb328ddeeac0718f07d57d6c799cfab3